### DISCUSSION TITLE
function as argument

### DISCUSSION THREAD

**Discussion by nico-robert** [apnadkarni/tcl-cffi#133] [General]
Hello @apnadkarni ,

I would like to use this C function :
`void Plot(const char* label_id, Point_getter getter, void* data, int count, Flags int);`

_Point_getter_  is defined like this:
`typedef void *(*Point_getter)(void* data, int idx, Point *pts);`

and _Point_ is a structure like this:
```
struct Point
{
    double x, y;
};
```
My goal is to set _Point_  when I call _Plot_ function, finally insert a pseudo function to set coordinates of my structure Point.

So far, what I tried :
```
cffi::Struct create Point {x float y float}
cffi::prototype function Point_getter void {data {pointer unsafe} idx int pts {pointer.Point unsafe}}
MYDLL function Plot void {label_id string getter pointer.Point_getter data {pointer nullok} count int Flags int}
```
I probably don’t understand the function, but for me it’s not a callback , rather I would think a cast of a pointer to a fonction, so that I can change my values in a TCL procedure.

If callback 
```
proc myTclProc {data idx pts} {
    foreach {x y} {200 300 400 500 600 700} {
        Point tonative $pts [list x $x y $y]
    }
}
set cb [cffi::callback new Point_getter myTclProc NULL] ; # I'm not sure for NULL value...
# C function
Plot "MyPlot" $cb NULL 3 0
```
I don’t know if I’m going in the right direction

**Comment by apnadkarni** [comment]
Assuming the `Point_getter` function (callback) will only be called while within a `Plot` function invocation, it looks ok in general with a few changes.

The `Point` struct fields in cffi should also be `double` as in the C version, not `float`.
```
cffi::Struct create Point {x double y double}
```
Since the callback return value is void, that last argument to `cffi::callback` (shown as NULL) need not be specified. 
```
set cb [cffi::callback new Point_getter myTclProc]
```

I think what the Point_getter function is supposed to return in a *single* point corresponding to the index, not an array of points.

```
proc myTclProc {data idx pPoint} {
     uplevel 1 coords values; # values is list in caller's context. Could be global also
     lassign [lindex $values $idx] x y; # Co-ords of idx'th point
     Point tonative $pPoint [list x $x y $y]
}
```

Then in the calling code,
```
 set coords {{200 300} {400 500} {600 700}}
 Plot "MyPlot" $cb NULL 3 0
```

I think that should work.

/Ashok

**Comment by apnadkarni** [comment]
As an aside, you could also pass the co-ordinates as a pointer to a native memory array via the `data` parameter. But that is a little more complicated in that the calling site will have to allocate an array of structs using cffi::memory or Point new. Using uplevel to reach caller's context seems easier.

**Reply by nico-robert** [reply]
I will try.

**Reply by nico-robert** [reply]
It doesn’t work but maybe the problem is with me. 
I have a question, if I have a function that is like that :

`PlotContext* Plot_CreateContext(void);` (`PlotContext` is a structure.)

In my function `Plot_CreateContext` returns a _pointer of structure_ ?
If Yes , Is it required to declare this structure  in `tcl-cffi `? Since finally I only return a pointer. Or I only have to declare it if I have to modify one or more parameters (fields)

**Comment by apnadkarni** [comment]
No, you do not have to declare the struct if you are only using pointers to a struct and not the struct value itself directly.

```
Plot_CreateContext pointer.PlotContext {}
```

will suffice.

If the Plot library is public, I can take a look as to why the callback does not work.

/Ashok

**Reply by nico-robert** [reply]
Hello @apnadkarni ,
yes this [library](https://github.com/sonoro1234/LuaJIT-ImGui) is public , but I have change the name's function to our discussion.
 Plot -> [ImPlot_PlotLineG ](https://github.com/cimgui/cimplot/blob/b7c6d712ecab51e72fcdd4d2b40ff79d226b3429/cimplot.h#L1534)
Point_getter ->[ImPlotPoint_getter ](https://github.com/cimgui/cimplot/blob/b7c6d712ecab51e72fcdd4d2b40ff79d226b3429/cimplot.h#L1531)
Point -> [ImPlotPoint](https://github.com/cimgui/cimplot/blob/b7c6d712ecab51e72fcdd4d2b40ff79d226b3429/cimplot.h#L343)

**Reply by nico-robert** [reply]
What I don’t understand is why even if my structure is not declared, I don’t have an error :
```
proc myTclProc {data idx pPoint} {
    ...
    dummyPT tonative $pPoint [list x $x y $y] ; # no error even if 'dummyPT' is not declared...
}
```

**Comment by apnadkarni** [comment]
I'll look at the ImPlot over the weekend.

Regarding dummyPt not showing an error, that is because the callback has no way to return an error to the invoking script  (since the return type is void). 

I also realized that pPoint is not a safe pointer so even if the dummyPT struct was defined, the `tonative` call would fail. You need to do something like

```
cffi::pointer safe $pPoint; # Mark as safe
dummyPt tonative $pPoint [list x $x y $y]
cffi::pointer dispose $pPoint; # Unmark as safe
```

Ideally there should have been a `tonative!` (like there is `fromnative!`) to deal with unsafe pointers but I seem to have overlooked that. Will need to add that so you do not need to do the safe/dispose dance above in the future.

I would suggest for debugging the callback, do something like

```
proc myTclProc {data idx pPoint} {
    set fd [open debug.log w]
    puts $fd "Entering myTclProc"
    try {
        cffi::pointer safe $pPoint; # Mark as safe
        dummyPt tonative $pPoint [list x $x y $y]
        cffi::pointer dispose $pPoint; # Unmark as safe
    } on error {result} {
        puts $fd $result
        error $result; # Rethrow error
    } finally {
        puts $fd "Leaving myTclProc"
        close $fd
    }
}
```

That might help debug the problem if you can see the actual error.

**Comment by apnadkarni** [comment]
Also taking a quick look at the ImPlot functions, it appears the definition 

```
cffi::prototype function Point_getter void {data {pointer unsafe} idx int pts {pointer.Point unsafe}}
```
is incorrect. The Point_getter function is defined as `void*`, not `void` so the definition should be 

```
cffi::prototype function Point_getter pointer {data {pointer unsafe} idx int pts {pointer.Point unsafe}}
```
I missed that in my earlier reading.

That means the callback is supposed to return a pointer. I am not sure what the pointer it returns is supposed to be. You might want to check the documentation on that.

I would be quite excited if you could get this ImGui integration working!

/Ashok

**Comment by apnadkarni** [comment]
And taking a closer look, the void/void* probably does not matter. The function return value is [thrown away](https://github.com/cimgui/cimplot/blob/b7c6d712ecab51e72fcdd4d2b40ff79d226b3429/cimplot.cpp#L2924) anyways.

**Comment by nico-robert** [comment]
Thanks , not present for the next two days , but I look at this when I come back to the office.

> I would be quite excited if you could get this ImGui integration working !

It's [done](https://github.com/ocornut/imgui/discussions/5576#discussioncomment-3483661) !! 

My goal is to integrate ImPlot into Tk frame but Imgui is very cool !  but there is a lot of functions to write.

**Comment by apnadkarni** [comment]
Is your repository public?

/Ashok

**Reply by nico-robert** [reply]
not yet , but I can share it with you if you want (from next Monday), if your answer is yes, I would prefer to share it via email.

**Comment by apnadkarni** [comment]
You only need to share the code if resolving the Point_getter callback issue is urgent. Else it can wait. You can send it to apnmbx-wits (yahoo.com)

**Reply by nico-robert** [reply]
thanks Ashok, it works , I only do a small change.
`uplevel 1 coords values; # values is list in caller's context. Could be global also`
by this :
`set values [uplevel 1 {set coords}]; # values is list in caller's context. Could be global also`

How did you know `$pPoint` was not safe pointer ?

**Comment by apnadkarni** [comment]
Arghh. I meant to write

```
upvar 1 coords values
```
and not
```
uplevel 1 coords values
```
What you are using will also work, but I think upvar will be more efficient.

Regarding `$pPoint` being unsafe, that's because it is a pointer that is "internal" to the ImPlot library and has not been seen by `CFFI` before and would not have been registered as a safe pointer. Generally speaking, this is true in most callbacks.

/Ashok

**Reply by nico-robert** [reply]
Okay, I can continue...
